Add cross-compilation support for Windows on Arm

richard-townsend-arm opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

Fixes https://github.com/nodejs/node/issues/32582. CC @richardlau @joaocgreis @jkunkee 

This PR fixes the Windows on Arm build of Node.js by applying all the necessary command line flags and workarounds, then adds cross-compilation support for the MSVC backend of node-gyp alongside supporting changes. The way this is implemented is that host GYP targets now end with _host.exe (e.g. mksnapshot_host.exe) and actions are re-written accordingly. In addition to the standard `node.sln`, a special `node_host.sln` is also generated with the host-only targets (this because MSBuild doesn't like building solutions with projects targeted at a mix of different architectures). 

**Current issues**
* Declaring support for cross-compilation means that some host-only projects are generated for x64/x64 too. The current solution is to compile host tools for all architectures, which increases build time and isn't strictly necessary (although a benefit is that somebody will definitely notice if they break cross-compilation in this setup). Possible changes may be needed to remove _host.exe tools before distribution. 
* node-gyp changes will need to be contributed upstream (?) and sent back down here. 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

richard-townsend-arm commented on vcbuild.bat:
```diff
@@ -374,7 +374,13 @@ if "%target%"=="Build" (
+  msbuild node_host.sln %msbcpu% /p:Configuration=%config% /p:Platform=x64 /clp:NoItemAndPropertyList;Verbosity=minimal /nologo %extra_msbuild_args%
+)
+echo node.sln %msbcpu% /t:%target% /p:Configuration=%config% /p:Platform=%msbplatform% /clp:NoItemAndPropertyList;Verbosity=minimal /nologo %extra_msbuild_args%
+msbuild node.sln %msbcpu% /t:%target% /p:Configuration=%config% /p:Platform=%msbplatform% /clp:NoItemAndPropertyList;Verbosity=minimal /nologo %extra_msbuild_args% /fl /flp:MSBuild.log;verbosity=diagnostic /detailedsummary
```
Should be resolved in the next push.

richard-townsend-arm commented on vcbuild.bat:
```diff
@@ -199,7 +199,7 @@ if defined config_flags     set configure_flags=%configure_flags% %config_flags%
 if defined openssl_no_asm   set configure_flags=%configure_flags% --openssl-no-asm
 if defined DEBUG_HELPER     set configure_flags=%configure_flags% --verbose
-if "%target_arch%"=="x86" if "%PROCESSOR_ARCHITECTURE%"=="AMD64" set configure_flags=%configure_flags% --no-cross-compiling
+if "%target_arch%"=="x86" if "%PROCESSOR_ARCHITECTURE%"=="AMD64" set configure_flags=%configure_flags%
```
Should be resolved in the next push.

richard-townsend-arm commented on vcbuild.bat:
```diff
@@ -374,7 +374,13 @@ if "%target%"=="Build" (
+if "%target_arch%" == "arm64" (
+  @rem Compile host-only tools first
+  echo msbuild node_host.sln %msbcpu% /p:Configuration=%config% /p:Platform=x64 /clp:NoItemAndPropertyList;Verbosity=minimal /nologo %extra_msbuild_args%
+  msbuild node_host.sln %msbcpu% /p:Configuration=%config% /p:Platform=x64 /clp:NoItemAndPropertyList;Verbosity=minimal /nologo %extra_msbuild_args%
```
Should be resolved in the next push.

richard-townsend-arm commented:
OK, looks like the MacOS CI failure is a flake, ready for review.

joaocgreis commented:
@richard-townsend-arm I pushed to your branch to make things easier, but I can change it back to 9effddd if you prefer. I changed the first line of the commit messages to follow the Node.js rules, and moved the vcbuild and configure changes to the last commit, so the middle commit is only gyp changes. We usually do this with commits that change dependencies, so it can be easily applied elsewhere or re-floated here if needed.

This looks good and works well for me. I can build x86, x64 and arm64 in my x64 machine. However, the `node_host.sln` file does not have any projects, it is only the minimal solution boilerplate. All the projects are built as part of the main solution, without errors. Do we really need a separate host sln file?

cclauss commented on tools/gyp/pylib/gyp/MSVSNew.py:
```diff
@@ -65,7 +65,7 @@ class MSVSFolder(MSVSSolutionEntry):
 
   def __init__(self, path, name = None, entries = None,
-               guid = None, items = None):
+               guid = None, items = None, projects_excluded_from_build=[]):
```
Eeekkk... Mutable default parameter https://docs.python-guide.org/writing/gotchas

cclauss commented on tools/gyp/pylib/gyp/MSVSNew.py:
```diff
@@ -172,7 +174,7 @@ class MSVSSolution(object):
 
   def __init__(self, path, version, entries=None, variants=None,
-               websiteProperties=True):
+               websiteProperties=True, projects_excluded_from_build=[]):
```
Eeekkk... Mutable default parameter https://docs.python-guide.org/writing/gotchas

richard-townsend-arm replied:
Interesting... I've been writing Python scripts for probably 9 years and I've never realised that this wasn't safe to do... learn something new every day I guess!

richard-townsend-arm replied:
These changes are (apparently) not needed, so I've reverted them in the next push.

cclauss commented on configure.py:
```diff
@@ -1040,7 +1040,10 @@ def configure_node(o):
   cross_compiling = (options.cross_compiling
                      if options.cross_compiling is not None
-                     else target_arch != host_arch)
+                     else (target_arch != host_arch
+                        and not (host_arch == "x64" and target_arch=="ia32")))
```
```suggestion
                     else host_arch not in (target_arch, "x64", "ia32"))
```

richard-townsend-arm replied:
Will address in the next push.

richard-townsend-arm commented:
> @richard-townsend-arm I pushed to your branch to make things easier, but I can change it back to [9effddd](https://github.com/nodejs/node/commit/9effddd9eca9bb9c5115abea667dd35de3ddae9b) if you prefer. I changed the first line of the commit messages to follow the Node.js rules, and moved the vcbuild and configure changes to the last commit, so the middle commit is only gyp changes. We usually do this with commits that change dependencies, so it can be easily applied elsewhere or re-floated here if needed.

Cool, that's great, thank you! (I guess I must have just mixed one of the commits up whilst rebasing). 

> This looks good and works well for me. I can build x86, x64 and arm64 in my x64 machine. However, the `node_host.sln` file does not have any projects, it is only the minimal solution boilerplate. All the projects are built as part of the main solution, without errors. Do we really need a separate host sln file?

So I've just double-checked this on my machine and... yeah, it seems to be working fine, without node_host.sln. I think that fixing the x86 build must have also fixed something about how the solution's set up, because I now see that it contains the expected mix of x64 and arm64 projects:
![image](https://user-images.githubusercontent.com/29576774/81232863-90f1ac80-8fed-11ea-8a6c-ba1f85d3cc00.png)

richard-townsend-arm commented:
Cool, so the next push will remove the node_host.sln file, and implement Christian's recommendations for improving node-gyp. Hopefully will get to it tomorrow.

richard-townsend-arm commented:
Cool, thanks @cclauss. I'm not authorized to merge. @joaocgreis, any further thoughts?

joaocgreis commented:
I've force pushed from c7e48545d62 to e71e4d365ba.

@richard-townsend-arm I didn't understand the change in configure, if I read it well if the host is x86 or x64 it never defaults to cross-compiling. I think this would be an issue for cross-compiling for the Raspberry Pi Linux, but apparently that job is not running at the moment in CI. I'd much rather leave the logic there as currently is and leave the line in vcbuild that it replaces.

I've rebased and added a fixup commit with my suggestion. The v8 commit is no longer necessary. Does this LGTY?

richard-townsend-arm commented:
Does look good to me, the commit you've pushed though is not building for me due to an unrelated error: 
```
C:\Users\rictow01\node\deps\v8\src\base\platform\platform-win32.cc(1407,5): error C2039: 'GetCurrentThreadStackLimits': is not a member of '`global namespace'' [C:\Us
ers\rictow01\node\tools\v8_gypfiles\v8_libbase.vcxproj]
C:\Users\rictow01\node\deps\v8\src\base\platform\platform-win32.cc(1407,32): error C3861: 'GetCurrentThreadStackLimits': identifier not found [C:\Users\rictow01\node\
tools\v8_gypfiles\v8_libbase.vcxproj]
```

... but that's separately fixable.

joaocgreis commented:
Rebased to get a fresh CI run (e71e4d365ba to 87f33894d13).

@richard-townsend-arm if this passes CI, I'll land this PR, as the new failure happens only on ARM64.

joaocgreis commented:
Landed in https://github.com/nodejs/node/compare/d093e788d18...88335512576

@richard-townsend-arm thanks for pushing this forward! The current failure building for ARM64 was caused by https://github.com/nodejs/node/commit/1d6adf7432defeb39b751a19c68335e8afb0d8ee#diff-4fa95c2da0a6e4a70f97811e38ef0ba8R1407 , the V8 update that landed while this PR was open. Do you know why that doesn't fail in V8 upstream (I believe they have ARM64 CI) and how to fix that here?

richard-townsend-arm commented:
I do super-vaguely remember fixing something like this on upstream V8, I’ll double-check and back port if necessary.

richard-townsend-arm commented:
Found it: https://chromium-review.googlesource.com/c/v8/v8/+/2124321

richard-townsend-arm commented:
Wait, sorry, that may not be quite the right fixup. I'll take a closer look in the next day or so.

richard-townsend-arm commented:
Actually, turned out to be something quite simple: this function is relatively new, so we need to define _WIN32_WINNT=0x0602 when compiling code involving GetCurrentThreadStackLimits. Will fix up and post a PR.
